### PR TITLE
Lazy-load action parameters

### PR DESCRIPTION
**Pull request by dshcherb**
In order to avoid an extra action-get call, this change introduces a
"params" property. "action-get" will return None if there are no
parameters defined for an action, however, in that case `.params` is unlikely to be used so an extra call will not matter.

Closes-Bug: #131

### CODE REVIEW

File: ops/charm.py [hunk 1]
```diff
@@ -18,8 +18,12 @@ def restore(self, snapshot):
...
             # This could only happen if the dev manually emits the action, or from a bug.
             raise RuntimeError('action event kind does not match current action')
-        # Params are loaded at restore rather than __init__ because the model is not available in __init__.
-        self.params = self.framework.model._backend.action_get()
+
+    @property
+    def params(self):
+        # NoneType will be returned if an action has no parameters defined so getattr is used instead of a check for NoneType.
+        self._params = getattr(self, '_params', self.framework.model._backend.action_get())
+        return self._params
```
Comment (jameinel) [inline ops/charm.py]: This syntax causes us to still evaluate _backend.action_get() on every lookup to self.params. Because functions passed to a function must be evaluated before the function is called.
What you want is something more like:
```
def __init__(self):
...
  self._params = None
...
def params(self):
  if self._params is None:
    self._params = self.framework.model._backend.action_get()
  return self._params
```

Also, this doesn't seem related to the rest of the patch, probably better to put it into its own patch.

Comment (dshcherb) [reply ops/charm.py]: The problem with ```if self._params is None:``` is that action-get will return `None` if there are no parameters defined. Thus on any `.params` access there will always be a hook tool execution.

I don't think this is terrible because people wouldn't normally access `params` if they know that an action does not accept any.

File: ops/charm.py [hunk 2]
```diff
@@ -9,17 +9,27 @@ class HookEvent(EventBase):
...
 
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._params = None
+
     def defer(self):
         raise RuntimeError('cannot defer action events')
 
     def restore(self, snapshot):
+        self._params = None
```
Comment (jameinel) [inline ops/charm.py]: why do we need to reset _params to None here.
Are params not part of the saved snapshot?

File: ops/charm.py [hunk 3]
```diff
@@ -9,17 +9,27 @@ class HookEvent(EventBase):
...
         event_action_name = self.handle.kind[:-len('_action')].replace('_', '-')
         if event_action_name != env_action_name:
             # This could only happen if the dev manually emits the action, or from a bug.
             raise RuntimeError('action event kind does not match current action')
-        # Params are loaded at restore rather than __init__ because the model is not available in __init__.
-        self.params = self.framework.model._backend.action_get()
+
+    @property
+    def params(self):
+        # NoneType will be returned if an action has no parameters defined but in that case this property is unlikely to be used.
```
Comment (jameinel) [inline ops/charm.py]: I think this is just "None will be returned".

However, if you call action-get, and we parse the result as JSON, don't we always end up with an Dict here, it will just be an empty dict if there are no parameters?

If we don't like returning None here, we can always do:
```
if self._params is None:
  self._params = ...
  if self._params is None:
    # no actual params, but avoid calling action-get again
    self._params = {}
return self._params
```

File: test/test_main.py [hunk 4]
```diff
@@ -206,8 +204,6 @@ def test_multiple_events_handled(self):
             'USE_ACTIONS': True,
         }))
 
-        fake_script(self, 'action-get', "echo '{}'")
```
Comment (jameinel) [inline test/test_main.py]: It seems surprising to me that the only test changes this causes is *removing* tests?
I suppose there may not be a visible change, as before you should be able to read params, and after you should be able to read the correct params.

File: ops/charm.py (hunk 2 above)
Comment (dshcherb) [reply ops/charm.py]: This is just a side-effect of how we emit events: they are persisted and reemitted. This results in a  strange behavior for `ActionEvents` because it happens for them as well while they cannot be manually deferred.

The problem with restoring event objects is that an `__init__` method is not called for unpickled objects and the framework calls `restore` on them:
https://github.com/canonical/operator/blob/ed093d5c3fa16454f87639550464376446e1713e/ops/framework.py#L490-L494

I don't think we need to store params into a unit-local state snapshot just to read them back right away. Similarly, we do not do anything like that with config on `config-changed` or relation data for relation data events.

File: ops/charm.py (hunk 3 above)
Comment (dshcherb) [reply ops/charm.py]: I added some test code into one of my charms (locally) before submitting and validated that it's not an empty dict.

It's a bit difficult to debug without `debug-hooks` support for actions but I got this by inserting an rpdb trace and running an action:

```
(Pdb) subprocess.check_output(['action-get', '--format=json'])
b'null\n'

(Pdb) type(json.loads(subprocess.check_output(['action-get', '--format=json']).decode('utf-8')))
<class 'NoneType'>
```

I think I'll go for your suggestion with an empty dict.

Comment (dshcherb) [reply ops/charm.py]: Interestingly, not using `--format=json` results in an empty dict being returned.

(Pdb) subprocess.check_output(['action-get']).decode('utf-8')                     
'{}\n'

Looks like a bug to me.

Comment (dshcherb) [reply ops/charm.py]: Filed here: https://bugs.launchpad.net/juju/+bug/1862026

Comment (dshcherb) [reply ops/charm.py]: Addressed in https://github.com/canonical/operator/pull/132/commits/61364bbf52b400d1a385f034046a4a84f7cdbf76

File: test/test_main.py (hunk 4 above)
Comment (dshcherb) [reply test/test_main.py]: Addressed in https://github.com/canonical/operator/pull/132/commits/61364bbf52b400d1a385f034046a4a84f7cdbf76

**Comment by dshcherb** [comment]
> This tastes somewhat like over-optimization. 
> Then, what's the most frequent action calling that we know about?

I was going for removing the unnecessary side-effect of calling a hook tool to be consistent with the rest of the code, less so for performance reasons.

> Aren't the parameters of a function already cached in the local agent by the time the function is called?

As far as I can see, Uniter retrieves an `ActionData` structure in order to fill out the relevant action environment variables:
https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/context/context.go#L739-L750

`ActionData` contains `Params` as a `map[string]interface{}` so 
https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/context/action.go#L14

`ActionData` is retrieved when a new `ActionRunner` is created
https://github.com/juju/juju/blob/juju-2.7.1/worker/uniter/runner/factory.go#L128-L133

It looks like you are right about the parameter caching.

**Comment by dshcherb** [comment]
Closing per https://github.com/canonical/operator/pull/132#pullrequestreview-354657870
